File: mycuration/src/main/java/com/phicdy/mycuration/presentation/presenter/CurationListPresenter.java
package com.phicdy.mycuration.presentation.presenter;

import android.support.annotation.NonNull;

import com.phicdy.mycuration.db.DatabaseAdapter;
import com.phicdy.mycuration.rss.Curation;
import com.phicdy.mycuration.presentation.view.CurationListView;

import java.util.ArrayList;

public class CurationListPresenter implements Presenter {

    private final DatabaseAdapter dbAdapter;
    private CurationListView view;

    public CurationListPresenter(DatabaseAdapter dbAdapter) {
        this.dbAdapter = dbAdapter;
    }

    public void setView(@NonNull CurationListView view) {
        this.view = view;
    }

    @Override
    public void create() {
    }

    @Override
    public void resume() {
        view.registerContextMenu();
        ArrayList<Curation> allCurations = dbAdapter.getAllCurations();
        view.initListBy(allCurations);
    }

    @Override
    public void pause() {
    }

    public void onCurationEditClicked(int curationId) {
        if (curationId < 0) return;
        view.startEditCurationActivity(curationId);
    }

    public void onCurationDeleteClicked(@NonNull Curation curation) {
        dbAdapter.deleteCuration(curation.getId());
        view.delete(curation);
    }

    public void activityCreated() {
        if (dbAdapter.getNumOfFeeds() == 0) {
            view.setNoRssTextToEmptyView();
        }
        view.setEmptyViewToList();
    }

    public int getCurationIdAt(int position) {
        if (position < 0 || position > view.size()) {
            return -1;
        }

        Curation curation = view.curationAt(position);
        if (curation == null) return -1;
        return curation.getId();
    }
}

File: mycuration/src/main/java/com/phicdy/mycuration/presentation/presenter/CurationListPresenter.kt
package com.phicdy.mycuration.presentation.presenter

import com.phicdy.mycuration.data.repository.CurationRepository
import com.phicdy.mycuration.data.repository.RssRepository
import com.phicdy.mycuration.data.repository.UnreadCountRepository
import com.phicdy.mycuration.domain.entity.Curation
import com.phicdy.mycuration.presentation.view.CurationItem
import com.phicdy.mycuration.presentation.view.CurationListView
import kotlinx.coroutines.coroutineScope
import java.util.ArrayList

class CurationListPresenter(private val view: CurationListView,
                            private val rssRepository: RssRepository,
                            private val curationRepository: CurationRepository,
                            private val unreadCountRepository: UnreadCountRepository) {
    private var allCurations: ArrayList<Curation> = arrayListOf()


    suspend fun resume() = coroutineScope {
        view.registerContextMenu()
        allCurations = curationRepository.getAllCurations()
        if (allCurations.isEmpty()) {
            view.hideRecyclerView()
            view.showEmptyView()
        } else {
            view.hideEmptyView()
            view.showRecyclerView()
            view.initListBy(allCurations)
        }
    }

    fun onCurationEditClicked(curationId: Int) {
        if (curationId < 0) return
        view.startEditCurationActivity(curationId)
    }

    suspend fun onCurationDeleteClicked(curation: Curation, size: Int) = coroutineScope {
        curationRepository.delete(curation.id)
        if (size == 1) {
            view.hideRecyclerView()
            view.showEmptyView()
        }
    }

    suspend fun activityCreated() = coroutineScope {
        if (rssRepository.getNumOfRss() == 0) {
            view.setNoRssTextToEmptyView()
        }
    }

    suspend fun getView(curation: Curation?, item: CurationItem) {
        curation?.let {
            item.setName(it.name)
            item.setCount(unreadCountRepository.getCurationCount(it.id).toString())
        }
    }
}
